### DISCUSSION TITLE
How are `bpf_bpf{eb,el}.o` files generated?

### DISCUSSION THREAD

**Discussion by hiqsociety** [cilium/ebpf#876] [Q&A] [answered]
for example xdp bpf_bpfeb.o <-- how is this file generated?

i would like to add libbpf_num_possible_cpus(), how do i do this?
```
int cpus = libbpf_num_possible_cpus();
```

**Comment by ti-mo** [comment, marked as answer]
The following `go:generate` statement invokes bpf2go when running `go generate` in the `examples/xdp` directory.
https://github.com/cilium/ebpf/blob/f17b01bb144a74559093d1699f5e03fac18070e8/examples/xdp/main.go#L24
This results in `bpf_bpf*.(g)o` files in the directory.

Regarding `libbpf_num_possible_cpus()`, please open another discussion with a bit more context about what you're trying to do.

**Reply by hiqsociety** [reply]
that's not true, if i remove the .o files found inside the directory, it fails to do. so it's not really running go generate to pre-generate the files

go build .

```
mv *.o /usr/local/src/
go build .
bpf_bpfel.go:119:12: pattern bpf_bpfel.o: no matching files found
```

**Reply by ti-mo** [reply]
> it's (`go build`) not really running go generate to pre-generate the files

Exactly, did I say it would?

> when running `go generate` in the `examples/xdp` directory

bpf2go invokes the compiler. This means `bpf_bpfel.o` is written when `go generate` is run.

**Reply by lmb** [reply]
Invoking `go build` doesn't invoke `go generate`, see https://go.dev/blog/generate.

**Comment by hiqsociety** [comment]
@lmb @ti-mo 

i rephrase then... 
1. how do i generate bpf_bpf{eb,el}.o files then?
2. how can i make xdp.c re-compile and show errors for compilation?

**Reply by lmb** [reply]
By invoking `go generate` in the correct folder as Timo and I have pointed out to you.
